### PR TITLE
Fixed the matmul function and its VJP.

### PR DESCRIPTION
**Pull request by eaplatanios**
The current implementation of `matmul` was implicitly reshaping arguments if necessary for the case of batch matrix multiplications. This causes unexpected behavior and it's better to have the user reshape as necessary before calling the function. Also, the VJP implementation was wrong due to these reshapes. The batch matmul op supports broadcasting and so I fixed the VJP, while also accounting for broadcasting.

### CODE REVIEW

**Comment by eaplatanios** [comment]
Is the description I added sufficient? If not, I can add more detail.

File: Sources/TensorFlow/Operators/Math.swift [hunk 1]
```diff
@@ -2114,16 +2114,11 @@ public func matmul<Scalar: Numeric>(
...
-    case (3..., 3...):
-        return Raw.batchMatMulV2(lhs, rhs, adjX: transposeA, adjY: transposeB)
-    case (2, 3...):
-        return Raw.batchMatMulV2(lhs.expandingShape(at: 1), rhs, adjX: transposeA, adjY: transposeB)
-    case (3..., 2):
-        return Raw.batchMatMulV2(lhs, rhs.expandingShape(at: 1), adjX: transposeA, adjY: transposeB)
-    default:
-        return Raw.matMul(lhs, rhs, transposeA: transposeA, transposeB: transposeB)
+    if lhs.rank > 2 || rhs.rank > 2 {
+      // TODO: Conjugate to make compatible with the adjoint.
```
Comment (rxwei) [inline Sources/TensorFlow/Operators/Math.swift]: File a bug for this?

Comment (eaplatanios) [reply Sources/TensorFlow/Operators/Math.swift]: I'll fix this as part of this PR. Give me a few minutes.

Comment (eaplatanios) [reply Sources/TensorFlow/Operators/Math.swift]: Actually, nvm this only matters if we support complex number types. Should I clarify this in the TODO alone or also file a bug?

Comment (rxwei) [reply Sources/TensorFlow/Operators/Math.swift]: Filing a bug sounds good.

Comment (eaplatanios) [reply Sources/TensorFlow/Operators/Math.swift]: Done.

**Pull request merged**